realip_test.go

```diff
@@ -75,10 +76,14 @@ func TestRealIP(t *testing.T) {
 			name:     "Has X-Forwarded-For",
 			request:  newRequest("", "", publicAddr1),
 			expected: publicAddr1,
+		}, {
+			name:     "Has X-Forwarded-For multiple IPs",
+			request:  newRequest("", "", fmt.Sprintf("%s,%s", localAddr, publicAddr1)),
+			expected: publicAddr1,
 		}, {
 			name:     "Has multiple X-Forwarded-For",
 			request:  newRequest("", "", localAddr, publicAddr1, publicAddr2),
-			expected: publicAddr2,
+			expected: publicAddr1,
```

Comment:
WHY should the first instance of X-Forwarded-For in the request be trusted? If I try to influence the behavior of a proxy between my client and my destination, wouldn't adding my own X-Forwarded-For header to my request be less trustworthy than any added by one after I left? What if I add two, do proxies between me and destination append to the first one (or all)?

Reply:
yup -- this is a known issue that it is spoofable.

Reply:
But to Dane’s point, if we know we control the last entry (or the last two) in the xff, they’re going to have a good value for the last public IP used on the path to connect to our next work.

Also this meme https://knowyourmeme.com/memes/good-luck-im-behind-7-proxies

Reply:
Yeah, I was discussing the differences of using `realip` and our haproxy-inserted headers in the long thread: https://udacity.slack.com/archives/GAVPMSL4A/p1533080802000176?thread_ts=1533067565.000178&cid=GAVPMSL4A

`realip` does use the first non-public IP as its algo (which means it is spoofable).

The author seems to accept that as a risk.

Using the last XFF (inserted by trusted infrastructure) is safer and is not spoofable.